File: ui/src/helpers/template.ts
/**
 * Copyright (c) 2020, The Facette Authors
 *
 * Licensed under the terms of the BSD 3-Clause License; a copy of the license
 * is available at: https://opensource.org/licenses/BSD-3-Clause
 */

interface TemplateNode {
    type: "text" | "variable";
    value: string;
}

function isVariableChar(c: string): boolean {
    return "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_".includes(c);
}

class Template {
    private s = "";

    private nodes: Array<TemplateNode> = [];

    public parse(text: string): void {
        this.s = text;
        this.nodes = [];

        while (this.s.length > 0) {
            const next = this.peek(2);
            if (next.length < 2) {
                this.nodes.push({type: "text", value: next});
                break;
            }

            if (next[0] === "$" && (isVariableChar(next[1]) || next[1] === "{")) {
                this.s = this.s.substr(1);
                this.nodes.push(this.readVariable());
            }

            this.nodes.push(this.readText());
        }
    }

    public render(data: Record<string, string>): string {
        return this.nodes
            .map((node: TemplateNode) => {
                switch (node.type) {
                    case "text":
                        return node.value;

                    case "variable":
                        return data[node.value] || "";
                }
            })
            .join("");
    }

    public variables(): Array<string> {
        return this.nodes.reduce((vars: Array<string>, node: TemplateNode) => {
            if (node.type === "variable" && !vars.includes(node.value)) {
                vars.push(node.value);
            }
            return vars;
        }, []);
    }

    private discard(): void {
        this.s = this.s.substr(1);
    }

    private peek(n: number): string {
        return this.s.substr(0, n);
    }

    private read(): string {
        const c: string = this.s.substr(0, 1);
        this.s = this.s.substr(1);
        return c;
    }

    private readText(): TemplateNode {
        let prev = "";
        let value = "";

        for (;;) {
            const c = this.read();
            if (c === "") {
                break;
            }

            if (c === "$" && prev !== "\\") {
                this.unread(c);

                const next = this.peek(2);
                if (isVariableChar(next[1]) || next[1] === "{") {
                    break;
                }

                this.discard();
            }

            value += c;
            prev = c;
        }

        return {type: "text", value};
    }

    private readVariable(): TemplateNode {
        let lbrace = false;
        let rbrace = false;

        const next = this.peek(1);
        if (next === "{") {
            lbrace = true;
            this.discard();
        }

        let value = "";

        for (;;) {
            const c = this.read();
            if (c === "") {
                break;
            }

            if (lbrace && c === "}") {
                rbrace = true;
                break;
            } else if (isVariableChar(c)) {
                value += c;
            } else {
                this.unread(c);
                break;
            }
        }

        if (lbrace && !rbrace) {
            throw Error("unbalanced brace");
        }

        return {type: "variable", value};
    }

    private unread(c: string): void {
        this.s = c + this.s;
    }
}

export function renderTemplate(text: string, data: Record<string, string>): string {
    const tmpl: Template = new Template();
    tmpl.parse(text);
    return tmpl.render(data);
}

export function parseVariables(text: string): Array<string> {
    const tmpl: Template = new Template();
    tmpl.parse(text);
    return tmpl.variables();
}

File: pkg/template/template.go
// Copyright (c) 2020, The Facette Authors
//
// Licensed under the terms of the BSD 3-Clause License; a copy of the license
// is available at: https://opensource.org/licenses/BSD-3-Clause

package template

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"sort"
	"strings"

	"facette.io/facette/pkg/set"
)

// Template is a representation of a parsed template.
type Template struct {
	r         *bufio.Reader
	pos, last position
	text      string
	nodes     []Node
}

// New creates a new template instance.
func New() *Template {
	return &Template{}
}

// Nodes returns the list of parsed template nodes.
func (t *Template) Nodes() []Node {
	return t.nodes
}

// Parse parses text as template data.
func (t *Template) Parse(text string) error {
	t.r = bufio.NewReader(strings.NewReader(text))
	t.pos = position{Line: 1, Char: 1}
	t.text = text
	t.nodes = []Node{}

	for {
		next, err := t.peek(2)
		if err == io.EOF {
			// Append remainder as text node
			if len(next) > 0 {
				t.nodes = append(t.nodes, &TextNode{string(next)})
			}

			break
		} else if err != nil {
			return err
		}

		var node Node

		if next[0] == '$' && (isVariableChar(next[1]) || next[1] == '{') {
			t.discard()

			node, err = t.readVariable()
			if err != nil {
				return err
			}

			t.nodes = append(t.nodes, node)

			continue
		}

		node, err = t.readText()
		if err != nil {
			return err
		}

		t.nodes = append(t.nodes, node)
	}

	return nil
}

// Variables returns the list of parsed variable names.
func (t *Template) Variables() []string {
	keys := set.New()

	for _, node := range t.nodes {
		v, ok := node.(*VariableNode)
		if ok {
			keys.Add(v.Name)
		}
	}

	result := set.StringSlice(keys)
	sort.Strings(result)

	return result
}

func (t *Template) discard() {
	_, _ = t.read()
}

func (t *Template) peek(n int) ([]byte, error) {
	return t.r.Peek(n)
}

func (t *Template) read() (byte, error) {
	b, err := t.r.ReadByte()
	if err == nil {
		t.last = t.pos

		if b == '\n' {
			t.pos.Line++
			t.pos.Char = 1
		} else {
			t.pos.Char++
		}
	}

	return b, err
}

func (t *Template) readText() (*TextNode, error) {
	var prev byte

	buf := bytes.NewBuffer(nil)

	for {
		c, err := t.read()
		if err == io.EOF {
			break
		} else if err != nil {
			return nil, err
		}

		if c == '$' && prev != '\\' {
			t.unread()

			next, err := t.peek(2)
			if err == nil && (isVariableChar(next[1]) || next[1] == '{') {
				break
			}

			t.discard()
		}

		buf.WriteByte(c)
		prev = c
	}

	return &TextNode{buf.String()}, nil
}

func (t *Template) readVariable() (*VariableNode, error) {
	var (
		lbrace, rbrace bool
		bracePos       position
	)

	next, err := t.peek(1)
	if err == nil && next[0] == '{' {
		lbrace, bracePos = true, t.pos
		t.discard()
	}

	buf := bytes.NewBuffer(nil)

loop:
	for {
		c, err := t.read()
		if err == io.EOF {
			break
		} else if err != nil {
			return nil, err
		}

		switch {
		case lbrace && c == '}':
			rbrace = true
			break loop

		case isVariableChar(c):
			buf.WriteByte(c)

		default:
			t.unread()
			break loop
		}
	}

	if lbrace && !rbrace {
		return nil, fmt.Errorf(`unbalanced brace at %s`, bracePos)
	}

	return &VariableNode{buf.String()}, nil
}

func (t *Template) unread() {
	err := t.r.UnreadByte()
	if err == nil && t.last.Line != 0 {
		t.pos = t.last
	}
}

type position struct {
	Line int
	Char int
}

func (p position) String() string {
	return fmt.Sprintf("%d:%d", p.Line, p.Char)
}

func isVariableChar(c byte) bool {
	return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9' || c == '_'
}
